support/agent/miner_agent.go

```diff
@@ -0,0 +1,338 @@
+	Worker        address.Address
+	IDAddress     address.Address
+	RobustAddress address.Address
+	PreCommits    uint64
+
+	operationSchedule *opQueue
+	deadlines         [miner.WPoStPeriodDeadlines][]bitfield.BitField
+	nextPrecommit     float64
+	nextSectorNumber  abi.SectorNumber
+	rnd               *rand.Rand
+}
+
+func NewMinerAgent(owner address.Address, worker address.Address, idAddress address.Address, robustAddress address.Address,
+	rnd *rand.Rand, config MinerAgentConfig,
+) *MinerAgent {
+	return &MinerAgent{
+		Config:        config,
+		Owner:         owner,
+		Worker:        worker,
+		IDAddress:     idAddress,
+		RobustAddress: robustAddress,
+
+		operationSchedule: &opQueue{},
+		nextPrecommit:     1.0 + precommitDelay(config.PrecommitRate, rnd), // next tick + random delay
+		rnd:               rnd,
+	}
+}
+
+func (ma *MinerAgent) Tick(v *vm.VM) ([]Message, error) {
+	var messages []Message
+
+	// start precommits, for now assume we have enough pledge funds
+	ma.nextPrecommit -= 1.0
+	for ma.nextPrecommit < 1.0 {
+		// go ahead and choose when we're going to activate this sector
+		sectorActivation := ma.sectorActivation(v.GetEpoch())
+		sectorNumber := ma.nextSectorNumber
+
+		messages = append(messages, ma.createPrecommit(v.GetEpoch(), sectorNumber, sectorActivation))
+
+		ma.nextPrecommit += precommitDelay(ma.Config.PrecommitRate, ma.rnd)
+		ma.nextSectorNumber++
+	}
+
+	// act on scheduled operations
+	for _, op := range ma.operationSchedule.PopOpsUntil(v.GetEpoch()) {
+		switch o := op.action.(type) {
+		case proveCommitAction:
+			messages = append(messages, ma.createProveCommit(o.sectorNumber))
+		case registerSectorAction:
+			err := ma.registerSector(v, o.sectorNumber)
+			if err != nil {
+				return nil, err
+			}
+		case proveDeadlineAction:
+			messages = append(messages, ma.proveDeadline(v, o.dlIdx))
+		}
+	}
+
+	return messages, nil
+}
+
+// proove sectors in deadline
+func (ma *MinerAgent) proveDeadline(v *vm.VM, dlIdx uint64) Message {
+	var partitions []miner.PoStPartition
+	for pIdx, bf := range ma.deadlines[dlIdx] {
+		if empty, err := bf.IsEmpty(); err != nil {
+			panic(err)
+		} else if !empty {
+			partitions = append(partitions, miner.PoStPartition{
+				Index:   uint64(pIdx),
+				Skipped: bitfield.New(),
+			})
+		}
+	}
+
+	postProofType, err := ma.Config.ProofType.RegisteredWindowPoStProof()
+	if err != nil {
+		panic(err)
+	}
+
+	params := miner.SubmitWindowedPoStParams{
+		Deadline:   dlIdx,
+		Partitions: partitions,
+		Proofs: []proof.PoStProof{{
+			PoStProof:  postProofType,
+			ProofBytes: []byte{},
+		}},
+		ChainCommitEpoch: v.GetEpoch() - 1,
+		ChainCommitRand:  []byte("not really random"),
+	}
+	return Message{
+		From:   ma.Worker,
+		To:     ma.RobustAddress,
+		Value:  big.Zero(),
+		Method: builtin.MethodsMiner.SubmitWindowedPoSt,
+		Params: &params,
+		ReturnHandler: func(v *vm.VM, _ Message, _ cbor.Marshaler) error {
+			return ma.scheduleNextProof(v, dlIdx)
```

Comment:
The callback mechanism (or something like it) is necessary for create miner because we need to capture its return value. For this an all other calls in miner agent it is unnecessary (so long as we assume that messages always succeed), so I'm moving all these handlers back into the main body and leaving the handler nil.